src/turbolinks/browser_adapter.coffee

```diff
@@ -14,6 +14,7 @@ class Turbolinks.BrowserAdapter
   visitStarted: (visit) ->
     visit.issueRequest()
     visit.changeHistory()
+    visit.goToSamePageAnchor()
     visit.loadCachedSnapshot()
```

Comment:
Calling `visit.goToSamePageAnchor` _and_ `visit.loadCachedSnapshot` on a restoration visit will result in `@adapter.visitRendered?(this)` being called twice. I think we may need a condition in either of those methods to prevent this.

Reply:
Conveniently enough, Visit's `render` method cancels any existing render calls, and so `@adapter.visitRendered?(this)` will never be called twice (the `visit.loadCachedSnapshot` call will override that in `visit.goToSamePageAnchor`) 🎉